### PR TITLE
IMPRO-1475:  remove accumulation from extrapolate

### PR DESCRIPTION
**Pull request by neilCrosswaite**
Acceptance
 * Reads the u/v-component single files created by IMPRO-1446 using existing utility functionality 
 * Produces rate forecasts at required (user-specified) lead times
 * Writes all rate forecasts to a single file with multiple validity times
    * Therefore no longer return accumulation

# As the CLI is now only returning one object, It will no longer be accumulating instead this role will be taken by a new accumulate CLI.
  
# Loading the cubelist

Loading is done with a constraint to specify the x and y.
# Refactor extrapolate

Change some of the error handling to inside the process function.
# Return 1 cube

Returns one cube after using merge_cube.

-Has to pop history to get it  to work.-

Can use improver merge cube.

 
# Tests
 * Removed accumulate tests
 * removed filename test as no the output file name is given

 h1. Testing the new cube match
```java
old0 = old/nowcast-extrapolate/extrapolate/kgo0.nc
old1 = old/nowcast-extrapolate/extrapolate/kgo1.nc
old2 = old/nowcast-extrapolate/extrapolate/kgo2.nc
new_kgo = new/nowcast-extrapolate/extrapolate/kgo.nc

old0 = load_cube(old0)
old1 = load_cube(old1)
old2 = load_cube(old2)
new*kgo = load_cube(new*kgo )

print((old0.data.data == new_kgo[0].data.data).all())
print((old1.data.data == new_kgo[1].data.data).all())
print((old2.data.data == new_kgo[2].data.data).all())


```
outputs
```python
True
True
True
```

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

File: lib/improver/cli/nowcast_extrapolate.py [hunk 1]
```diff
@@ -31,17 +31,15 @@
 # POSSIBILITY OF SUCH DAMAGE.
 """Script to extrapolate input data given advection velocity fields."""
 
-import os
-
 import iris
 import numpy as np
 from iris import Constraint
+from iris.cube import CubeList
```
Comment (benfitzpatrick) [inline lib/improver/cli/nowcast_extrapolate.py]: Codacy Issue found: [Unused CubeList imported from iris.cube](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4499606)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

File: tests/improver-nowcast-extrapolate/02-basic.bats [hunk 2]
```diff
@@ -33,38 +33,25 @@
...
 @test "extrapolate basic" {
   improver_check_skip_acceptance
-  KGO0="nowcast-extrapolate/extrapolate/kgo0.nc"
-  KGO1="nowcast-extrapolate/extrapolate/kgo1.nc"
-  KGO2="nowcast-extrapolate/extrapolate/kgo2.nc"
+  KGO="nowcast-extrapolate/extrapolate/kgo.nc"
 
-  UCOMP="$IMPROVER_ACC_TEST_DIR/nowcast-optical-flow/basic/ucomp_kgo.nc"
-  VCOMP="$IMPROVER_ACC_TEST_DIR/nowcast-optical-flow/basic/vcomp_kgo.nc"
+  UVCOMP="$IMPROVER_ACC_TEST_DIR/nowcast-optical-flow/basic/kgo.nc"
```
Comment (bayliffe) [inline tests/improver-nowcast-extrapolate/02-basic.bats]: This file has been recreated in your new test data but matches the data already in the archive.

Comment (neilCrosswaite) [reply tests/improver-nowcast-extrapolate/02-basic.bats]: fixed.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

File: lib/improver/cli/nowcast_extrapolate.py [hunk 3]
```diff
@@ -169,18 +169,17 @@ def process(input_cube, u_cube, v_cube, speed_cube, direction_cube,
         ValueError:
             can either use s_cube and d_cube or u_cube and v_cube.
             Therefore: (s and d)⊕(u and v)
```
Comment (bayliffe) [inline lib/improver/cli/nowcast_extrapolate.py]: speed_cube and direction_cube

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

File: lib/improver/cli/nowcast_extrapolate.py (hunk 3 above)
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_extrapolate.py]: Done

File: lib/improver/cli/nowcast_extrapolate.py [hunk 4]
```diff
@@ -219,85 +160,34 @@ def process(input_cube, u_cube, v_cube, speed_cube, direction_cube,
...
-        fraction, _ = np.modf(max_lead_time / accumulation_fidelity)
-        if fraction != 0:
-            msg = ("The specified lead_time_interval ({}) is not cleanly "
-                   "divisible by the specified accumulation_fidelity ({}). As "
-                   "a result the lead_time_interval cannot be constructed from"
-                   " accumulation cubes at this fidelity.")
-            raise ValueError(msg.format(lead_time_interval,
-                                        accumulation_fidelity))
-        time_interval = accumulation_fidelity
+    if (speed_cube and direction_cube) and not (u_cube or v_cube):
```
Comment (gavinevans) [inline lib/improver/cli/nowcast_extrapolate.py]: Do the previous `if` tests mean that this `if` test can be reduced to `if speed_cube and direction_cube`? As far as I can tell, I would have thought that, as a result of your previous tests, you already know that if `speed_cube` and `direction_cube` are not None, then `u_cube` and `v_cube` will be None.

Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_extrapolate.py]: fixed

File: lib/improver/cli/nowcast_extrapolate.py [hunk 5]
```diff
@@ -99,90 +81,49 @@ def main(argv=None):
...
-        "--accumulation_period", type=int, default=15,
-        help="The period over which the accumulation is calculated (mins). "
-        "Only full accumulation periods will be computed. At lead times "
-        "that are shorter than the accumulation period, no accumulation "
-        "output will be produced.")
-    accumulation_args.add_argument(
-        "--accumulation_units", type=str, default='m',
-        help="Desired units in which the accumulations should be expressed,"
-        "e.g. mm")
+    parser.add_argument("--u_and_v_filepath", type=str, help="Path to u and v"
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_extrapolate.py]: Done

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity decreasing per file
==============================
+ lib/improver/cli/nowcast_extrapolate.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4499606&bid=15169072)

**Pull request merged**